### ISSUE TITLE
policy that only allows putobject doesn't work without explicit deny

### ISSUE THREAD

**Issue by rgevaert** [minio/minio#11038]
Hi,

I want to create a policy that only allows to upload objects in an a bucket that has the same name as the username.
This policy should be enough I would think:
```
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:PutObject"
   ],
   "Resource": [
    "arn:aws:s3:::${aws:username}/*"
   ]
  }
 ]
}
```

however when I applied to a user it still lets me delete objects.

Extending the policy with an explicit deny does what I want to do:
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": [
        "s3:PutObject"
      ],
      "Effect": "Allow",
      "Resource": [
        "arn:aws:s3:::${aws:username}/*"
      ],
      "Sid": "BucketUploadForUser"
    },
    {
      "Sid": "DenyDelete",
      "Effect": "Deny",
      "Action": "s3:DeleteObject",
      "Resource": [
        "arn:aws:s3:::${aws:username}/*"
      ]
    }
  ]
}
```

I find it very strange that I need to explicitly deny it. This is very strange from a security perspective. Is this working as intended?

thanks!

**Comment by harshavardhana** [comment]
@rgevaert how are you testing deletes ? Through browser?

**Comment by rgevaert** [comment]
Hi no with MC RM.

Op za 5 dec. 2020 18:18 schreef Harshavardhana <notifications@github.com>:

**Comment by harshavardhana** [comment]
> Hi no with MC RM.

`mc rm` would have failed @rgevaert `mc` doesn't show an error that's all.

**Comment by harshavardhana** [comment]
Here is how I tested it 

```
~  MC_HOST_test=https://test:test123456@play.minio.io/ mc rm test/test/hosts 
mc: <ERROR> Failed to remove `test/test/hosts`. Insufficient permissions to access this file `https://play.minio.io/test/hosts`

~ mc admin user info play/ test
AccessKey: test
Status: enabled
PolicyName: write
MemberOf: 

~ mc admin policy info play/ write
{
 "Version": "2012-10-17",
 "Statement": [
  {
   "Effect": "Allow",
   "Action": [
    "s3:PutObject"
   ],
   "Resource": [
    "arn:aws:s3:::${aws:username}/*"
   ]
  }
 ]
}
```

So looks like this is working fine

**Comment by rgevaert** [comment]
Thanks @harshavardhana ! This are the commands I did yesterday. The user belonged to a group and I put the policy on the group. I have now tried it again and indeed it works like you say.
Looking at my outputs again from yesterday I think I found the issue:

```
❯ mc admin group info play upload-only
Group: upload-only
Status: enabled
Policy: upload-only-username
Members: user1,user2

❯ mc admin user info play user1
AccessKey: user1
Status: enabled
PolicyName: user
MemberOf: upload-only

```

In testing things I had set also a `user` policy on the user too. I can't look it it anymore but I think that one will have had more access.

Sorry for wasting your time!

Rudy
